File: android/feature/catalog/src/test/kotlin/HomeShelvesTest.kt
package catalog

import model.Kind
import model.MediaSet
import model.Progress
import model.Watched
import model.WatchSnapshot
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * The start page answers two questions: what was already underway, and what
 * turned up. A library is added to over years, so the order these rows are
 * in is the whole of what they are for — get it wrong and the page is a
 * handful of shelves with fewer things on them.
 */
class HomeShelvesTest {

    @Test
    fun filmsAreNewestFirst() {
        val rows = homeRowsOf(shelvesOf(listOf(film("Old", at = 100), film("New", at = 900))), WatchSnapshot.Empty)

        val films = rows.single { it.title == "Latest movies" }.entries()
        assertEquals(listOf("New", "Old"), films.map { (it as Entry.Film).set.title })
    }

    /**
     * A series still being uploaded keeps its place: it is dated by its
     * newest episode, not by the one that happened to arrive first.
     */
    @Test
    fun aShowIsDatedByItsNewestEpisode() {
        val rows = homeRowsOf(
            shelvesOf(
                listOf(
                    episode("Started long ago", episode = 1, at = 100),
                    episode("Started long ago", episode = 2, at = 5_000),
                    episode("Arrived whole", episode = 1, at = 900),
                ),
            ),
            WatchSnapshot.Empty,
        )

        val shows = rows.single { it.title == "Latest series" }.entries()
        assertEquals(
            listOf("Started long ago", "Arrived whole"),
            shows.map { (it as Entry.Collection).name },
            "the show that gained an episode most recently leads",
        )
    }

    @Test
    fun aRowHoldsSixAndLeavesTheRestToItsShelf() {
        val many = (1..20).map { film("Film $it", at = it.toLong()) }

        val row = homeRowsOf(shelvesOf(many), WatchSnapshot.Empty).single()
        assertEquals(HOME_ROW_LIMIT, row.entries().size)
        assertEquals(20, row.total, "the heading counts the whole shelf")
        assertEquals("Film 20", (row.entries().first() as Entry.Film).set.title, "newest leads")
    }

    /** The row names the shelf it is a window onto, so "See all" can open it. */
    @Test
    fun everyRowNamesTheShelfBehindIt() {
        val rows = homeRowsOf(shelvesOf(listOf(film("Alien", at = 1), lesson("Steuerkurs", at = 2))), WatchSnapshot.Empty)

        assertEquals(setOf("Movies", "Tutorials"), rows.map { it.seeAll }.toSet())
        assertTrue(rows.all { it.title.startsWith("Latest ") }, "rows borrow the shelf's own word")
    }

    /** An empty library has nothing to say, and says nothing. */
    @Test
    fun anEmptyLibraryProducesNoRows() {
        assertTrue(homeRowsOf(emptyList(), WatchSnapshot.Empty).isEmpty())
    }

    /**
     * An index that recorded no arrival time still lists. It sorts last
     * rather than first, because an unknown date is not the beginning of
     * time as far as a viewer is concerned.
     */
    @Test
    fun aSetWithNoArrivalTimeStillAppears() {
        val rows = homeRowsOf(shelvesOf(listOf(film("Undated", at = 0), film("Dated", at = 10))), WatchSnapshot.Empty)

        val films = rows.single().entries().map { (it as Entry.Film).set.title }
        assertEquals(listOf("Dated", "Undated"), films)
        assertNull(films.firstOrNull { it == "Missing" })
    }

    /** A film half-watched leads Continue, and carries the whole library's started count. */
    @Test
    fun aStartedFilmLeadsContinue() {
        val alien = film("Alien", at = 1)
        val watch = WatchSnapshot(
            progress = listOf(Progress(alien.setId, at = 1_800.0, duration = 3_600.0, updatedAt = 10)),
            watched = emptyList(),
            watchlist = emptyList(),
            kids = emptyList(),
            collections = emptyList(),
        )

        val rows = homeRowsOf(shelvesOf(listOf(alien)), watch)
        val continueRow = rows.single { it.title == "Continue" }
        val cards = (continueRow.content as RowContent.Sets).cards
        assertEquals(listOf(alien.setId), cards.map { it.set.setId })
        assertEquals(1, continueRow.total)
        assertNull(continueRow.seeAll, "Continue has no tab to open until the kept-shelves phase")
    }

    /** A finished episode offers the next one under Next up, whose "See all" opens Series. */
    @Test
    fun aFinishedEpisodeOffersNextUp() {
        val e1 = episode("Show", episode = 1, at = 1)
        val e2 = episode("Show", episode = 2, at = 1)
        val watch = WatchSnapshot(
            progress = emptyList(),
            watched = listOf(Watched(e1.setId, finishedAt = 100)),
            watchlist = emptyList(),
            kids = emptyList(),
            collections = emptyList(),
        )

        val rows = homeRowsOf(shelvesOf(listOf(e1, e2)), watch)
        val nextUpRow = rows.single { it.title == "Next up" }
        val cards = (nextUpRow.content as RowContent.Sets).cards
        assertEquals(listOf(e2.setId), cards.map { it.set.setId })
        assertEquals("Series", nextUpRow.seeAll)
    }
}

private fun HomeRow.entries(): List<Entry> = (content as RowContent.Entries).entries

private fun film(title: String, at: Long) = set(Kind.MOVIE, title, addedAt = at)

private fun episode(show: String, episode: Int, at: Long) =
    set(Kind.EPISODE, "Episode $episode of $show", show = show, season = 1, episodeFirst = episode, addedAt = at)

private fun lesson(course: String, at: Long) =
    set(Kind.TUTORIAL, "Lektion 1", show = course, path = "Grundlagen", addedAt = at)

private fun set(
    kind: Kind,
    title: String,
    show: String? = null,
    path: String? = null,
    season: Int? = null,
    episodeFirst: Int? = null,
    addedAt: Long = 0,
) = MediaSet(
    setId = "$kind-$show-$title",
    kind = kind,
    title = title,
    show = show,
    chapter = null,
    path = path,
    season = season,
    episodeFirst = episodeFirst,
    episodeLast = episodeFirst,
    year = null,
    durationSecs = null,
    posterPath = null,
    totalBytes = 0,
    addedAt = addedAt,
)

File: web/test/home-shelves.test.ts
/**
 * What the start page shows.
 *
 * Two facts make the page: what arrived recently, and what was underway.
 * The second is the one with rules — which episode of a show to offer when
 * the viewer has a position in one and has finished another — and every one
 * of those rules is asserted here, because none of them is visible from a
 * screenshot until it is wrong.
 */

import { describe, expect, test } from "bun:test";
import { groupLibrary, type CatalogSet } from "../public/lib/library.js";
import { homeShelves } from "../public/lib/catalog/home-shelves.js";

const set = (over: Record<string, unknown> = {}): CatalogSet => ({
  setId: `01SET${Math.random().toString(36).slice(2, 10).toUpperCase()}`,
  kind: "movie",
  title: "A Title",
  show: null,
  path: null,
  chap: null,
  season: null,
  episode: null,
  year: null,
  container: "mp4",
  vcodec: "h264",
  acodec: "aac",
  duration: 3600,
  total: 1000,
  partCount: 1,
  addedAt: 1_000,
  ...over,
});

/** An episode of `show`, numbered, with a stable id to record state against. */
const episode = (show: string, season: number, number: number, over: Record<string, unknown> = {}) =>
  set({
    setId: `${show}-s${season}e${number}`.toUpperCase().replace(/[^A-Z0-9]/g, ""),
    kind: "ep",
    show,
    season,
    episode: String(number),
    title: `${show} ${season}x${number}`,
    ...over,
  });

/** A position half way through an hour, which `resumeAt` accepts. */
const halfway = (setId: string, updatedAt: number) => ({
  setId,
  at: 1800,
  duration: 3600,
  updatedAt,
});

/** The page's inputs, from a flat list of sets and some watch state. */
function shelvesOf(
  sets: CatalogSet[],
  state: { progress?: ReturnType<typeof halfway>[]; watched?: Record<string, number> } = {},
  limit?: number,
) {
  const watched = state.watched ?? {};
  return homeShelves({
    library: groupLibrary(sets),
    byId: new Map(sets.map((one) => [one.setId, one])),
    // The store hands these over newest first; so does this.
    progress: [...(state.progress ?? [])].sort((a, b) => b.updatedAt - a.updatedAt),
    watchedAt: (setId: string) => watched[setId] ?? null,
    limit,
  });
}

describe("what arrived recently", () => {
  test("films are newest first, whatever the shelf sorted them by", () => {
    const shelves = shelvesOf([
      set({ title: "Old", addedAt: 1_000 }),
      set({ title: "New", addedAt: 3_000 }),
      set({ title: "Middle", addedAt: 2_000 }),
    ]);

    expect(shelves.latestMovies.map((one) => one.title)).toEqual(["New", "Middle", "Old"]);
  });

  test("a show ranks by its newest episode, not by its first", () => {
    // The point of the rule: a series still being uploaded keeps its place,
    // and one whose first episode is ancient is not buried by that.
    const shelves = shelvesOf([
      episode("Long Running", 1, 1, { addedAt: 100 }),
      episode("Long Running", 9, 9, { addedAt: 9_000 }),
      episode("One Season", 1, 1, { addedAt: 5_000 }),
    ]);

    expect(shelves.latestSeries.map((one) => one.name)).toEqual(["Long Running", "One Season"]);
  });

  test("courses rank the same way, by their newest lesson", () => {
    const shelves = shelvesOf([
      set({ kind: "tut", show: "Java", title: "L1", addedAt: 10 }),
      set({ kind: "tut", show: "Rust", title: "L1", addedAt: 20 }),
      set({ kind: "tut", show: "Java", title: "L2", addedAt: 30 }),
    ]);

    expect(shelves.latestCourses.map((one) => one.name)).toEqual(["Java", "Rust"]);
  });

  test("a row holds no more than the limit", () => {
    const many = Array.from({ length: 12 }, (_, n) => set({ title: `Film ${n}`, addedAt: n }));

    expect(shelvesOf(many).latestMovies).toHaveLength(6);
    // The heading still counts the whole shelf, not the six on the row.
    expect(shelvesOf(many).totals.latestMovies).toBe(12);
    expect(shelvesOf(many, {}, 3).latestMovies).toHaveLength(3);
  });

  test("a library with nothing in it gives rows that are empty, not absent", () => {
    // The view decides what to draw; the model never answers `undefined`.
    const shelves = shelvesOf([]);

    expect(shelves.latestMovies).toEqual([]);
    expect(shelves.nextUp).toEqual([]);
    expect(shelves.continues).toEqual([]);
  });
});

describe("what is underway", () => {
  const show = [episode("Show", 1, 1), episode("Show", 1, 2), episode("Show", 1, 3)];

  test("a show nobody has touched is not on the row", () => {
    expect(shelvesOf(show).nextUp).toEqual([]);
  });

  test("a part-watched episode is offered, as itself", () => {
    const shelves = shelvesOf(show, { progress: [halfway(show[1]!.setId, 5_000)] });

    expect(shelves.nextUp).toHaveLength(1);
    expect(shelves.nextUp[0]!.set.setId).toBe(show[1]!.setId);
    expect(shelves.nextUp[0]!.resume).toBe(true);
  });

  test("a finished episode offers the one after it", () => {
    const shelves = shelvesOf(show, { watched: { [show[0]!.setId]: 5_000 } });

    expect(shelves.nextUp[0]!.set.setId).toBe(show[1]!.setId);
    expect(shelves.nextUp[0]!.resume).toBe(false);
  });

  test("being part-way through beats what follows a finished episode", () => {
    // Finished 1 and then started 2: the viewer is in the middle of 2, and
    // offering 3 would skip what they are actually watching.
    const shelves = shelvesOf(show, {
      watched: { [show[0]!.setId]: 5_000 },
      progress: [halfway(show[1]!.setId, 6_000)],
    });

    expect(shelves.nextUp[0]!.set.setId).toBe(show[1]!.setId);
    expect(shelves.nextUp[0]!.resume).toBe(true);
  });

  test("an episode already watched out of order is skipped, not offered again", () => {
    const shelves = shelvesOf(show, {
      watched: { [show[0]!.setId]: 5_000, [show[1]!.setId]: 1_000 },
    });

    expect(shelves.nextUp[0]!.set.setId).toBe(show[2]!.setId);
  });

  test("a show with nothing left leaves the row", () => {
    const shelves = shelvesOf(show, {
      watched: { [show[0]!.setId]: 1, [show[1]!.setId]: 2, [show[2]!.setId]: 3 },
    });

    expect(shelves.nextUp).toEqual([]);
  });

  test("a glance is not being underway", () => {
    // Twenty seconds into an hour is looking at it, which `resumeAt`
    // refuses — and a show with no other mark has nothing to caption.
    const shelves = shelvesOf(show, {
      progress: [{ setId: show[0]!.setId, at: 20, duration: 3600, updatedAt: 5_000 }],
    });

    expect(shelves.nextUp).toEqual([]);
  });

  test("shows come in the order they were last watched", () => {
    const other = [episode("Other", 1, 1), episode("Other", 1, 2)];
    const shelves = shelvesOf([...show, ...other], {
      progress: [halfway(show[0]!.setId, 1_000)],
      // Finished more recently than the other show was left part-way.
      watched: { [other[0]!.setId]: 9_000 },
    });

    expect(shelves.nextUp.map((entry) => entry.collection.name)).toEqual(["Other", "Show"]);
  });

  test("a completion ranks a show even though it cleared the position", () => {
    // The case the dated `watched` row exists for: finishing an episode
    // erases its progress, and without the date this show would rank last.
    const other = [episode("Other", 1, 1), episode("Other", 1, 2)];
    const shelves = shelvesOf([...show, ...other], {
      progress: [halfway(other[0]!.setId, 8_000)],
      watched: { [show[0]!.setId]: 9_000 },
    });

    expect(shelves.nextUp[0]!.collection.name).toBe("Show");
  });

  test("courses are underway the same way series are", () => {
    const lessons = [
      set({ kind: "tut", show: "Java", title: "L1", path: null }),
      set({ kind: "tut", show: "Java", title: "L2", path: null }),
    ];
    const shelves = shelvesOf(lessons, { watched: { [lessons[0]!.setId]: 5_000 } });

    expect(shelves.nextUp).toHaveLength(1);
    expect(shelves.nextUp[0]!.set.title).toBe("L2");
  });
});

describe("the same title never appears twice", () => {
  const show = [episode("Show", 1, 1), episode("Show", 1, 2)];

  test("an episode under Next up is not also under Continue", () => {
    const film = set({ title: "A Film" });
    const shelves = shelvesOf([...show, film], {
      progress: [halfway(show[0]!.setId, 9_000), halfway(film.setId, 8_000)],
    });

    expect(shelves.nextUp[0]!.set.setId).toBe(show[0]!.setId);
    expect(shelves.continues.map((one) => one.setId)).toEqual([film.setId]);
    // The count still agrees with the Continue shelf, which lists both.
    expect(shelves.totals.continues).toBe(2);
    expect(shelves.totals.nextUp).toBe(1);
  });

  test("an episode Next up had no room for still reaches Continue", () => {
    // Nothing underway silently disappears from the page: the subtraction
    // is against what is *shown*, not against everything that qualified.
    const other = [episode("Other", 1, 1), episode("Other", 1, 2)];
    const shelves = shelvesOf(
      [...show, ...other],
      {
        progress: [halfway(show[0]!.setId, 9_000), halfway(other[0]!.setId, 8_000)],
      },
      1,
    );

    expect(shelves.nextUp.map((entry) => entry.set.setId)).toEqual([show[0]!.setId]);
    expect(shelves.continues.map((one) => one.setId)).toEqual([other[0]!.setId]);
  });

  test("Continue is most recent first and holds only resumable titles", () => {
    const older = set({ title: "Older" });
    const newer = set({ title: "Newer" });
    const glanced = set({ title: "Glanced" });
    const shelves = shelvesOf([older, newer, glanced], {
      progress: [
        halfway(older.setId, 1_000),
        halfway(newer.setId, 2_000),
        { setId: glanced.setId, at: 5, duration: 3600, updatedAt: 3_000 },
      ],
    });

    expect(shelves.continues.map((one) => one.title)).toEqual(["Newer", "Older"]);
  });

  test("a position against a title the catalog no longer holds is dropped", () => {
    const film = set({ title: "A Film" });
    const shelves = shelvesOf([film], {
      progress: [halfway("01GONEFROMTHECATALOG", 9_000), halfway(film.setId, 1_000)],
    });

    expect(shelves.continues.map((one) => one.title)).toEqual(["A Film"]);
  });
});
